### DISCUSSION TITLE
[ZK Stack] - How to aggregate more tx per batch

### DISCUSSION THREAD

**Discussion by Archethect** [zkSync-Community-Hub/zksync-developers#572] [ZK Stack] [answered]
### Module Affected

- [ ] System contracts (era-system-contracts)
- [ ] Smart contracts (era-contracts)
- [ ] zkEVM (era-zk_evm)
- [X] Server (zksync-era)
- [X] Prover (era-boojum)
- [ ] Fee withdrawer (era-fee-withdrawer)
- [ ] Block Explorer (clock-explorer)
- [ ] Portal - Wallet + Bridge (dapp-portal)
- [ ] Other (please specify below)

### Rust Version

1.75.0

### Operating System

- [X] Ubuntu
- [ ] MacOS
- [ ] Windows
- [ ] Other (Please specify below)

### Issue Description

We are currently experiencing a bot farm executing multiple tx/second on our chain.
How can we aggregate more tx in our batches to avoid an excessive amount of almost empty batches?
I know that ZKSync Era seals batches roughly every minute. On our side currently the batches seal every couple of seconds which results in batches that can't be executed in time on L1 to keep up with the L2 transactions.

Which setting can we change to aggregate more tx per batch?

### Expected Behavior

Batches contain more trsnsactions and seal slower (every X minutes instead of every X seconds).

### Current Behavior

Batches seal every 5-10 seconds resulting in batches that can't be executed fast enough on L1 to keep up with the L2 tx.

### Additional Details

Tx and batches can be monitored here: https://albireo-explorer.playfi.ai

### Prior Research

- [X] Yes, I searched for similar issues and did not find any matching ones.

**Comment by Archethect** [comment]
I guess one of these settings need to be changed:

```CHAIN_STATE_KEEPER_TRANSACTION_SLOTS=250
CHAIN_STATE_KEEPER_MAX_ALLOWED_L2_TX_GAS_LIMIT=4000000000
CHAIN_STATE_KEEPER_BLOCK_COMMIT_DEADLINE_MS=2500
CHAIN_STATE_KEEPER_MINIBLOCK_COMMIT_DEADLINE_MS=1000
CHAIN_STATE_KEEPER_MINIBLOCK_SEAL_QUEUE_CAPACITY=10
CHAIN_STATE_KEEPER_MAX_SINGLE_TX_GAS=6000000
CHAIN_STATE_KEEPER_CLOSE_BLOCK_AT_GEOMETRY_PERCENTAGE=0.95
CHAIN_STATE_KEEPER_CLOSE_BLOCK_AT_ETH_PARAMS_PERCENTAGE=0.95
CHAIN_STATE_KEEPER_CLOSE_BLOCK_AT_GAS_PERCENTAGE=0.95
CHAIN_STATE_KEEPER_REJECT_TX_AT_GEOMETRY_PERCENTAGE=0.95
CHAIN_STATE_KEEPER_REJECT_TX_AT_ETH_PARAMS_PERCENTAGE=0.95
CHAIN_STATE_KEEPER_REJECT_TX_AT_GAS_PERCENTAGE=0.95
CHAIN_STATE_KEEPER_MINIMAL_L2_GAS_PRICE=100000000
CHAIN_STATE_KEEPER_COMPUTE_OVERHEAD_PART=0
CHAIN_STATE_KEEPER_PUBDATA_OVERHEAD_PART=1
CHAIN_STATE_KEEPER_BATCH_OVERHEAD_L1_GAS=800000
CHAIN_STATE_KEEPER_MAX_GAS_PER_BATCH=200000000
CHAIN_STATE_KEEPER_MAX_PUBDATA_PER_BATCH=100000
CHAIN_STATE_KEEPER_FEE_MODEL_VERSION=V1
CHAIN_STATE_KEEPER_VALIDATION_COMPUTATIONAL_GAS_LIMIT=300000
CHAIN_STATE_KEEPER_SAVE_CALL_TRACES=true
CHAIN_STATE_KEEPER_VIRTUAL_BLOCKS_INTERVAL=1
CHAIN_STATE_KEEPER_VIRTUAL_BLOCKS_PER_MINIBLOCK=1
CHAIN_STATE_KEEPER_UPLOAD_WITNESS_INPUTS_TO_GCS=false

**Reply by qiwenzhou** [reply, marked as answer]
First, I am not from zksync company, just a dev from community. 
Second, to aggregate more txs, CHAIN_STATE_KEEPER_BLOCK_COMMIT_DEADLINE_MS is set to be larger to have more blocks before sealing batch. CHAIN_STATE_KEEPER_TRANSACTION_SLOTS is set to be max txs number in every batch.

**Reply by Archethect** [reply]
Thanks @qiwenzhou. That seemed to do the trick!

**Reply by qiwenzhou** [reply]
Glad to be of help

**Reply by sachanayush47** [reply]
> First, I am not from zksync company, just a dev from community. Second, to aggregate more txs, CHAIN_STATE_KEEPER_BLOCK_COMMIT_DEADLINE_MS is set to be larger to have more blocks before sealing batch. CHAIN_STATE_KEEPER_TRANSACTION_SLOTS is set to be max txs number in every batch.

Where I should make these changes?
Upon searching the given keywords I found this file `zksync-era/core/lib/env_config/src/chain.rs`, but Im not sure if this is the appropriate file to make changes

**Reply by qiwenzhou** [reply]
> > First, I am not from zksync company, just a dev from community. Second, to aggregate more txs, CHAIN_STATE_KEEPER_BLOCK_COMMIT_DEADLINE_MS is set to be larger to have more blocks before sealing batch. CHAIN_STATE_KEEPER_TRANSACTION_SLOTS is set to be max txs number in every batch.
> 
> Where I should make these changes? Upon searching the given keywords I found this file `zksync-era/core/lib/env_config/src/chain.rs`, but Im not sure if this is the appropriate file to make changes

under zksync_era/etc/env/ you can find {your projoct}.env. That is where you make those changes
